### ISSUE TITLE
Remove Wikibase extension from all OSM wikis

### ISSUE THREAD

**Issue by Firefishy** [openstreetmap/operations#764]
The `Wikibase` extension causes us endless compatibility issues with Mediawiki version. 

It is overly complex to install and requires specific knowledge to manage.

I propose we remove the extension and restore the wiki back to standard functionality as best as possible.

**Comment by not-my-profile** [comment]
I think the two main data consumers of the OSM wiki wikibase data are the [iD editor](https://github.com/openstreetmap/iD) and [Sophox](https://github.com/Sophox/sophox).
The iD editor displays descriptions and images from the data items as follows (when you click on the `i`):

![image](https://user-images.githubusercontent.com/73739153/198863958-e7da58ce-cbde-49c5-ac1e-dacb89958ded.png)

and the :pen: links to the respective data item (in this case [Q98](https://wiki.openstreetmap.org/wiki/Item:Q98)). Sophox on the other hand has apparently not been importing updates to the data items in nearly 1.5 years, see https://github.com/Sophox/sophox/issues/27. Note that taginfo does not use the Wikibase data at all (and was not planning on supporting it either), see https://github.com/taginfo/taginfo/issues/248.

While I am a big fan of structured data and think that we should embrace it as much as possible for the documentation of our tagging conventions, I do think that Wikibase is a poor fit for our use case because it requires data items to have numeric IDs ... which does make sense for Wikidata but not for tags because keys and values are already unique identifiers. Unfortunately Wikibase really seems to be primarily targeted at Wikidata. Another example being that we cannot disable the label fields for data items and they're often mistakenly translated as well.

Lastly the current state of the data items is quite a mess because there are absolutely no mechanisms in place to synchronize template data with the data items ... it's all done manually. And when you create a new page on the wiki there's no data item created for the page so many tag pages don't even have data items.

So with these caveats in mind I guess it makes sense to abandon Wikibase if it is a maintenance burden. I think we'd definitely need to migrate the tag description translations somewhere else ... IMHO regular wiki pages are definitely not an option ... we'd need some service with a similarly userfriendly translation UI.

I think ideally we could switch to some Wikibase alternative that better fit our use case but I don't think there is any.

**Comment by mmd-osm** [comment]
> there are absolutely no mechanisms in place to synchronize template data with the data items 

IIRC, @nyurik used a bot called Yurikbot to do this kind of synchronization, but for some reason,it's no longer active: https://wiki.openstreetmap.org/wiki/User_talk:Yurik#Yurikbot_2

**Comment by AndrewHain** [comment]
I would strongly be opposed to losing the knowledge base built up in the data items. Despite being hampered by opposition from one prolific and vocal wiki contributor and a loss of interest form Yuri, they are in better shape than the difficult-to-maintain arguments to the wiki templates. I also used data items effectively to [stop the wiki rendere from choking](https://wiki.openstreetmap.org/wiki/Talk:Map_features#Performance_rewrite_(and_support_for_Translate_extension)).

If Wikibase is too hard to manage the most sensible alternative would be to remove tag documentation from the wiki entirely and move it to a new custom website.

**Comment by SomeoneElseOSM** [comment]
> I would strongly be opposed to losing the knowledge base built up in the data items.

@AndrewHain Can you explain what this knowledge base is and how people can use it, to add to what @not-my-profile  wrote above?  This probably isn't best done here, but an OSM diary entry might be a good place for it.

**Comment by Firefishy** [comment]
I appreciate the extra detail being provided.

**Comment by Firefishy** [comment]
This ticket should not be considered a 100% decision point... It is more a cry of pain needing a solution 😜

**Comment by 1ec5** [comment]
> The Wikibase extension causes us endless compatibility issues with Mediawiki version.

For the benefit of those coming into this discussion, can you elaborate on the specific issues you anticipate going forward if we keep Wikibase around? Are these compatibility issues with MediaWiki itself or with other extensions we have installed? Do we have a staging environment to test our MediaWiki configuration with Wikibase before deployment, to catch issues before they disrupt ordinary wiki usage?

> I think the two main data consumers of the OSM wiki wikibase data are the [iD editor](https://github.com/openstreetmap/iD) and [Sophox](https://github.com/Sophox/sophox).

In addition, the wiki itself depends on data items in various ways. For example, when you search for a key that doesn’t have an article yet or land on a 404 page for that key, the wiki displays an infobox synthesized from that key’s data item. If it’s a compound key, the 404 page also [includes a breakdown of the key](https://wiki.openstreetmap.org/wiki/Talk:Wiki/Archive_9#Enhanced_not-found_messages) based on the data item of each component.

This functionality was added in response to concern about having to maintain articles in each language about arbitrary combinations of key components. An alternative would be to populate these articles using a bot, but the descriptions would have to be maintained somewhere more machine-readable than infoboxes in wikitext, essentially reinventing data items.

iD consumes data items through the MediaWiki API. It’s a public API, so we don’t know for sure what other software (QA tools?) rely on it for descriptions, images, or statements about which element types are valid for a given tag. A breaking change of this magnitude needs to be discussed broadly, in a similar manner as if taginfo or Nominatim were to introduce a breaking change in its API.

> Sophox on the other hand has apparently not been importing updates to the data items in nearly 1.5 years, see https://github.com/Sophox/sophox/issues/27.

This is inaccurate. Sophox/sophox#27 is tracking a bug that Sophox has been omitting many OSM elements, mostly nodes. It is also behind on ingesting OSM planet data overall. However, it’s up-to-date with respect to data items.

**Comment by flacombe** [comment]
Dear all,

I second comments against loosing the knowledge and practices that came with wikibase.
All we need is to improve it, not abandon it.

> Lastly the current state of the data items is quite a mess because there are absolutely no mechanisms in place to synchronize template data with the data items ... it's all done manually

And there shouldn't be any. Tags structured data should be put once in data items with no redundancy in wiki.
Redundancy is a temporary bridge, waiting for a more robust integration. It seems this had last long enough to be a valid point to abandon the whole architecture, this is not good.
Let's finish the work prior to challenge it in the middle of the journey.

UX on wikibase editor has been criticized for years with not so much involvement to make it better.
This is huge work, but :
How can we achieve moving the whole documentation on a brand new system (we don't know yet) OSM community will manage on its own, if we're unable to improve existing and common tools?

**Comment by tomhughes** [comment]
The problem with wikibase is basically as Firefishy said, that it requires significant configuration to make it work and that configuration frequently changes and is poorly documented.

As far as I know very people people (other than wikidata obviously) actually run wikibase and as such it's not well tested unless you happen to be running the exact same bleeding edge version as wikidata - also they run in a separate mediawiki instance while we try and run it all in the one instance.

The result is that it frequently breaks when we upgrade mediawiki and I have to ask Yuri for help and he has to ask the wikidata people and they mostly shrug and say they have no idea if the release version we're using will work and it's all just a shit show.

**Comment by not-my-profile** [comment]
>https://github.com/Sophox/sophox/issues/27 is tracking a bug that Sophox has been omitting many OSM elements

Ah yeah my bad ... I got confused by the different Sophox issues.

> However, it’s up-to-date with respect to data items, modulo a potential problem with dropping older data items:

I do not think that it's up to date with regards to data items. E.g. [this query](https://sophox.org/#select%20%2a%20where%20%7B%0A%20%20osmd%3AQ22096%20%3Ffoo%20%3Fbar.%0A%7D) yields nothing despite [the data item](https://wiki.openstreetmap.org/wiki/Item:Q22096) existing for 11 days.

> The main alternative is [Semantic MediaWiki](https://www.semantic-mediawiki.org/wiki/Semantic_MediaWiki),

I do have experience with SMW and am certain that it's not an improvement over Wikibase.

> All we need is to improve it, not abandon it.

I don't think that we can address the limitations of the Wikibase software (such as numeric ids being required).

> Tags structured data should be put once in data items with no redundancy in wiki.

This does not work because taginfo does not support it and the taginfo maintainer is not interested in supporting it. Hence the redundancy. And I can understand their point because the current data item situation really is a mess.

> How can we achieve moving the whole documentation on a brand new system (we don't know yet) OSM community will manage on its own, if we're unable to improve existing and common tools?

Migrating data to a compatible system would be easy ... it's just that there is no such suitable system afaik. Some system written specifically with OSM in mind could work much better for us. You cannot adapt a system if your use case is not in the scope of the project ... unless you fork the project but Wikibase is way too complicated to be forked.

Even if we do not see a solution to the situation right now, I think it's good for us to have this discussion.

**Comment by flacombe** [comment]
I got the point about maintenance and updates. I'll try to forward it to Wikimedia people we know at OSM France chapter.

> As far as I know very people people (other than wikidata obviously) actually run wikibase

Wikidata + OSM is still twice more instances than OSM itself on a custom software.

> I don't think that we can address the limitations of the Wikibase software (such as numeric ids being required).
How the numeric id is a limitation?

Can't the items be reach through P19 https://wiki.openstreetmap.org/wiki/Property:P19 ?

> This does not work because taginfo does not support it and the taginfo maintainer is not interested in supporting it. Hence the redundancy. And I can understand their point because the current data item situation really is a mess.

How could that change if we don't manage to change ourselves?

> Some system written specifically with OSM in mind could work much better for us

Shouldn't we address needs on core osm.org website or API instead?
No one could provide our API or core software but that would be clever to take advantage from Wikipedia experience.

> You cannot adapt a system if your use case is not in the scope of the project ...

Come on, can't we talk to be part of this scope?

**Comment by not-my-profile** [comment]
> How the numeric id is a limitation?

We do not need numeric ids but we have to use them ... that's a limitation.

>Can't the items be reach through P19?

Yes of course but it's very much an unnecessary layer of indirection / source of confusion/inconvenience.

>Shouldn't we address needs on core osm.org website or API instead?

This is not an either or scenario. I do consider documentation to be a very important topic.

> Come on, can't we talk to be part of this scope?

Have a look at [T202676](https://phabricator.wikimedia.org/T202676), where a comparatively very small (but useful) change was rejected. Numeric ids on the other hand are fundamental to Wikibase, this is not something that can be easily changed and I am very certain that this won't change ... so this is not really a talking matter.

**Comment by fititnt** [comment]
@tomhughes @Firefishy There's any technical reasons to the Wikibase central point **not** be a dedicated place, let's say, https://base.openstreetmap.org ? Not saying about the name, but mostly to simplify infra, because is possible to both not be same instance.

This doesn't solve 100% of the obvious challenge (which I've already aware by the issues with upgrading the OpenStreetMap Wiki; others please read at least this https://github.com/openstreetmap/operations/issues/760#issuecomment-1280802385) however even Wikipedia have an dedicated Wikibase (https://www.wikidata.org/) installation to isolate the chaos from all the wikis. This somewhat means that while the OSM use of Wikibase is actually more likely to be harder to maintain (including backup size, scalability issues, etc) than Wikipedia's approach.

And like I wrote on the Wiki:Talk, I agree that this issue should be raised as affecting the smoother upgrades on the Wiki.

**Comment by not-my-profile** [comment]
Wikipedia has a dedicated Wikibase instance because there are many different Wikipedia instances (e.g. en.wikipedia.org, de.wikipedia.org, es.wikipedia.org etc.). I am pretty sure that moving the Wikibase instance of the OSM to a different domain would not solve or improve anything. The wikibase client extension would still need to be kept in sync with the wikibase repository extension on the different instance.

**Comment by 1ec5** [comment]
> The result is that it frequently breaks when we upgrade mediawiki and I have to ask Yuri for help and he has to ask the wikidata people and they mostly shrug and say they have no idea if the release version we're using will work and it's all just a shit show.

There is an uphill climb for third-party installations of MediaWiki, but for what it’s worth, Wikimedia’s sister projects have often faced the same hurdles when it comes to maintaining extensions that Wikipedia doesn’t use (like [Proofread Page](https://www.mediawiki.org/wiki/Extension:Proofread_Page) for Wikisource). Local chapters such as Wikimedia Deutschland and Wikimedia Italia have been instrumental to supporting non-Wikipedia use cases, so I applaud @flacombe’s outreach to Wikimedia France.

Perhaps the OSMF should explore joining the [Wikibase Stakeholder Group](https://wbstakeholder.group/), which sponsors improvements to Wikibase to make it more reusable independently of WMF infrastructure. Judging from the member list, we aren’t alone in needing this support structure. In parallel, the [Wikibase Community User Group](https://meta.wikimedia.org/wiki/Wikibase_Community_User_Group) advocates for increased reusability and is open to individual participation.

> however even Wikipedia have an dedicated Wikibase (https://www.wikidata.org/) installation to isolate the chaos from all the wikis

The reason Wikidata exists is to centralize data from myriad Wikimedia wikis, not to isolate Wikibase from those wikis. Quite the contrary: all the other wikis have [Wikibase Client](https://www.mediawiki.org/wiki/Extension:Wikibase_Client) installed, and Wikimedia Commons has even overlaid another Wikibase instance to track [structured data](https://commons.wikimedia.org/wiki/Special:MyLanguage/Commons:Structured_data) about media files.

Separating Wikibase from the main wiki wouldn’t by itself reduce maintenance costs. There are hosted solutions like [wikibase.cloud](https://www.wikibase.cloud/), but even if we manage to outsource Wikibase, I think the OSM Wiki would still need to install Wikibase Client and get existing data consumers to transition over to the new instance. Being a client of a separate Wikimedia Commons media repository has been its own headache…

**Comment by Firefishy** [comment]
I'd feel a lot more comfortable if we were to split off `wikibase` to a separate instance of mediawiki, splitting it from wiki.openstreetmap.org.

**Comment by nyurik** [comment]
@Firefishy there is very little point in splitting wikibase because it consists of two parts -- server and client, and the client must live in the osm wiki to be of any reasonable use -- which means you would actually increase the number of problems (cross-site integration) rather than keeping it relatively simple.  This is exactly how it is done with all other wiki installations that chose to use wikibase. It was also mentioned by a few other people above.

**Comment by tomhughes** [comment]
We haven't even gone to 1.38 yet (when I looked a week or so back 1.39 wasn't out yet) because I got burnt so badly by the last upgrade that I can't face trying to do it again. We're still on 1.37 for the main wiki.

**Comment by tomhughes** [comment]
In fact 1.39 is still not out, but we do want to move to it once it is, for PHP 8 support.

**Comment by matkoniecz** [comment]
> And there shouldn't be any. Tags structured data should be put once in data items with no redundancy in wiki.

Note that it should be proposed, discussed and agreed on OSM Wiki before actually doing this.

> when you search for a key that doesn’t have an article yet or land on a 404 page for that key, the wiki displays an infobox synthesized from that key’s data item. If it’s a compound key, the 404 page also [includes a breakdown of the key](https://wiki.openstreetmap.org/wiki/Talk:Wiki/Archive_9#Enhanced_not-found_messages) based on the data item of each component.

that is quite useful! Though all other mentioned uses are not actually requiring data items or wikibase

**Comment by nyurik** [comment]
@tomhughes was the previous migrate issue came about because the upgrade was done in production without testing it on the staging servers first?  As far as I know, that was the only incident with Wikibase. Or is that a different issues?

**Comment by tomhughes** [comment]
We don't have a staging server...

The recent incident that happened with the 1.37 upgrade led to https://github.com/openstreetmap/chef/commit/ec0bc46d1275fc39a116736fc40372b0f6784fd0 as my first attempt to fix it and then a few days later after I got you involved https://github.com/openstreetmap/chef/commit/586ac89854fe37ae4d9a0b8dcda6797535cafb49 which was what finally got it working again.

I think there was at least one previous occasion where I had to get your help after an upgrade but I can't recall the details.

**Comment by matkoniecz** [comment]
> If Wikibase is too hard to manage the most sensible alternative would be to remove tag documentation from the wiki entirely and move it to a new custom website.

Wiki worked before data items were added and will continue working (with some very minor functionality missing and some other functionality replaced by other solution) in case of Wikibase being disabled.

I get that some people invested a lot of effort into data items but they are not irreplaceable. Please do not present OSM Wiki shutdown as a possible consequence. That is simply misleading.

> I think we'd definitely need to migrate the tag description translations somewhere else ... IMHO regular wiki pages are definitely not an option ... 

Regular wiki pages and preset translations fulfil well this needs, data item translations are of a very dubious use.

Quality is dubious, existing uses can be easily replaced by extracting data from infoboxes and serving that as an API (maybe there would be greater delay in updates, what would be worth lower risk of definitions being changed without any explanation and oversight)

> iD consumes data items through the MediaWiki API. It’s a public API, so we don’t know for sure what other software (QA tools?) rely on it for descriptions, images, or statements about which element types are valid for a given tag. A breaking change of this magnitude needs to be discussed broadly, in a similar manner as if taginfo or Nominatim were to introduce a breaking change in its API.

Definitely. I posted info at OSM Wiki to notify interested people.

**Comment by flacombe** [comment]
> Note that it should be proposed, discussed and agreed on OSM Wiki before actually doing this.

This was said from a state of the art IT perspective.
Every change will be discussed prior to be released.

> existing uses can be easily replaced by extracting data from infoboxes and serving that as an API

This doesn't solve the inconsistency in translations issue the wikibase is intended to solve.
Here is the only place I know where crawling inconsistent text is preferred over a structured database (wikibase or other one).
Current taginfo already offers this API but fails in restoring the consistency (it should be restored by people with appropriate tools actually).
This is not said to blame maintainers who achieve great things but the messy architecture we collectively encourage.

**Comment by 1ec5** [comment]
> We don't have a staging server...

Upgrading in production without staging first sounds like a deeper problem. Today the obstacles are Wikibase and MultiMaps; tomorrow it may be something else in our configuration unrelated to these extensions. A staging server won’t magically fix extension incompatibilities, but without this important workflow, any issue becomes a fire drill, and the mere threat of such fire drills leads to nihilist proposals.

> Regular wiki pages and preset translations fulfil well this needs, data item translations are of a very dubious use.

How can we be confident that this is not a minority opinion?

> Definitely. I posted info at OSM Wiki to notify interested people.

You’re more optimistic than I am that every consumer of the OSM Wiki’s MediaWiki API instance follows the Talk:Wiki page or this repository. Given [Hyrum’s law](https://xkcd.com/1172/), it’s inevitable that something would be broken in any transition of data items off the wiki, no matter who we contact; it’s only a question of the extent to which we care. But as a courtesy, there probably should be a more visible heads-up on one of the mailing lists, ideally focused on  the immediate technical issues – @Firefishy’s plea for help – rather than running a victory lap around data items.

**Comment by tomhughes** [comment]
Well are we expected to run a second staging server for every service we run?

I don't really think it's practical to get rid of wikidata at this point. I do wish I'd never let Yuri persuade me to add it though.

**Comment by matkoniecz** [comment]
> This doesn't solve the inconsistency in translations issue the wikibase is intended to solve.

Maybe wikibase was intended to solve it, but is not solving it at all but makes it worse (no edit description, ineffective watchlisting result in problem being even worse)

> structured database

infoboxes are also structured and wikibase is inferior in many ways, not only in how much it is annoying for sysadmins but also in interface quality and broken watchlisting

**Comment by flacombe** [comment]
> wikibase is inferior in many ways, not only in how much it is annoying for sysadmins but also in interface quality and broken watchlisting

Glad we didn't get rid of any annoying OSM interface this way prior to improve it.
Rome wasn't build in one day, they're waiting our pull requests.

**Comment by fititnt** [comment]
Fact: **the _Wiki people_ and _the Wikibase people_ have near irreconcilable points of view about the structure of the same information**. Something similar happens on Wikidata vs Wikipedia. Likely no consensus will be archived on a regular basis.

So, another (and maybe actually more relevant reason) to split is not just infra scale better, but to allow different self governance. What is called "data items" actually are used not just by the wiki, but by others, such as developers who would use it for translations of terms in their interfaces. Also, the OSM Wikibase currently is highly underutilized.

I don't think people agree to give up. So this is less the issue. **However, for the _Wikibase people_ here: why a dedicated place under some OSM domain would be bad?**

The Wiki can have client enabled (and worst case, NGinx/Apache can redirect, so the URLs still work; but this Wiki decision) however this change could allow decisions that affect the semantic not be _a pet project_ of the policy of the Wiki. Not even Wikipedia does this. Also, the OpenStreetMap Ops give some signal that splitting in different domains already could be an option.

**Comment by 1ec5** [comment]
> I don't think people agree to give up. So this is less the issue. However, for the Wikibase people here: why a dedicated place under some OSM domain would be bad?

I appreciate your effort at finding middle ground. However, so far, @Firefishy has floated the idea of not even installing Wikibase Client on the OSM Wiki: https://github.com/openstreetmap/operations/issues/764#issuecomment-1297253138. This is more extreme than Wikipedia, which you cite as precedent. It doesn’t answer the question of how the wiki will continue to source information that it’s currently getting from data items (such as in infoboxes and Map Features). Wikibase Client is not guaranteed to be any less of a maintenance burden when upgrading, as it’s designed to be compatible with Wikibase.

Jettisoning extensions is not a straightforward long-term strategy for streamlining MediaWiki upgrades, because there are tradeoffs and it won’t stop at Wikibase. (Remember when we accidentally took down the wiki by installing Scribunto?) “Wikibase people” have been willing to acknowledge that Wikibase comes with maintenance overhead, but conversely there should be acknowledgement that there is a legitimate need for what the status quo has been providing, for all its warts.

What we need to move forward is a clear explanation of what will break if we upgrade to MediaWiki 1.39 in #760. This will enable us to identify solutions or seek help from experts. Ideally we’d be running a staging server, but even a local dry run should be able to turn up whatever is causing us lost sleep. Then we can open issues about it in the relevant bug trackers. The energy currently devoted to this issue would be much effective if channeled at the Wikibase developers, once we have something concrete to point to. Our problems aren’t really on their radar yet. Let’s save the blue-sky ideas about a novel replacement until we’ve exhausted our options.

**Comment by nyurik** [comment]
Thanks @1ec5, I have been thinking along the same lines.

After reading through all the comments, it sounds like the main problem is not the wikibase itself, but devops scaling.

The devops demand has grown -- more software, with some becoming unmaintained -- but the (amazing!) current devops team really needs help, yet doesn't have an organized way to scale up.

I would love to help. I have been in devops with the "keys to the castle" in a 3000+ ppl company for many years, so I do have some experience, but I currently don't see any path for me to help - there is no clear way to on-board new devops volunteers, there is no process to try to set up staging systems so that the upgrade becomes more predictable and peaceful, there is no way to try new approaches -- e.g. dockers / kubernetes / maybe even cloud-based deployments with terraform+helm -- all those things enjoyed by many startups and fortune500s alike.

I am not saying these techs is what's needed. What's needed is a way to reliably onboard new people, and allow them to be productive. Otherwise something will continue crashing, and relying on just a few dedicated individuals is not any reasonable long-term effort should do.

**Comment by dpriskorn** [comment]
> > wikibase is inferior in many ways, not only in how much it is annoying for sysadmins but also in interface quality and broken watchlisting
> 
> 
> 
> Glad we didn't get rid of any annoying OSM interface this way prior to improve it.
> 
> Rome wasn't build in one day, they're waiting for our pull requests.

I agree the Wikibase UI is still lacking now ten years into the project. I would very much encourage this community to clearly report any UX problems we experience. 
To my knowledge a core problem with Wikibase development is a total lack of usability testing. We could do a couple of recordings with user testing and report back our findings to WMDE.

**Comment by matkoniecz** [comment]
>  Rome wasn't build in one day, they're waiting for our pull requests.

Not really. See https://phabricator.wikimedia.org/T202676 (can anyone share a success story?)

**Comment by GreenReaper** [comment]
Not *all* third-party installs use a 'combined' Wikibase. It might not be a bad idea to split it into a separate wiki, as at least Repository-specific failures would not directly break the main wiki, and there are ways to reduce the pain ([shared tables](https://mediawiki.org/wiki/Manual:Shared_database) or [CentralAuth](https://mediawiki.org/wiki/Extension:CentralAuth)); however it might also increase the maintenance workload, and moreover I am not sure it is even practical at this point.

Not only would [certain tables need to be moved](https://mediawiki.org/wiki/Wikibase/Schema), the actual entities are [stored as JSON documents in the main database tables](https://doc.wikimedia.org/Wikibase/master/php/docs_storage_entities.html) (with [secondary storage](https://doc.wikimedia.org/Wikibase/master/php/docs_storage_terms.html) in the aforementioned tables for Wikibase's needs). They are at least in their own namespaces, but still - it'd involve slicing up tables, and I don't know that anyone has done it before. It might be easier to just export and import relevant namespaces, with a new WDQS as well. [The concept URI would also change](https://addshore.com/2019/11/changing-the-concept-uri-of-an-existing-wikibase-with-data/), although the external impact of this may be limited if you do not have many external users of such data.

It you *do* split it, I'd advise staying within the same database cluster, otherwise integration options such as [client editing](https://mediawiki.org/wiki/Wikidata_Bridge) will be limited.

---

As for the question of staging, a separate environment might be a first start, either on the same server (but separate domains and PHP-FPM pools) or perhaps somewhere in the cloud. The free instances on offer from [Oracle](https://docs.oracle.com/en-us/iaas/Content/FreeTier/freetier_topic-Always_Free_Resources.htm) or [Google](https://cloud.google.com/free/docs/free-cloud-features#free-tier) may be plenty, since you don't need to have it process serious traffic.

**Comment by angoca** [comment]
My point of view on removing the Data items part from the wiki could affect many people who had worked on this part, correcting many things, improving texts, etc. They will see this removal as if their long working hours were for nothing, and they will probably be discouraged from continuing contributing to the Wiki or even to OSM. The wiki is our knowledge base, and it has grown in different dimensions thanks to the collaboration of many volunteers; we cannot cut one part just because of some technical issues.

As a database expert, I see the values in the KeyDescription, and ValueDescription templates in regular wiki pages for the same key/value in different languages as redundant, with many inconsistencies among them. And we know redundancy without synchronization is a bad practice that leads to incorrect values; therefore, the mappers unknown the absolute truth. I have modified [thousands of regular wiki pages (literally)](https://wiki.openstreetmap.org/wiki/Special:Contributions/Angoca) trying to standardize just one element of the tags, the "status," which is different in some languages for many entries. And this is a never-ending task because there is nothing centralized. In this case, what is the purpose of having the chance to put a [different "status" value other than the one in English](https://taginfo.openstreetmap.org/keys/highway#wiki)? The status is unique for a key or value, no matter the language on the wiki page is written. And this problem applies to many options in the KeyDescription and ValueDescription.

I perceive the data items like a normalized database with a single possible value for a specific element. Instead of removing the data item functionality, we should do the opposite and encourage its usage. I know the learning curve is complex, and there are not too many people working on it, but we can write more articles about how to contribute and maintain this part to welcome new volunteers in this area. We also must remember that with the increasing number of new tags and translations, maintaining the regular wiki articles will be even more complicated, and they will be even more disorganized. Instead, if we start to migrate and standardize values in different languages to a unique value in the data item, the maintenance will be easier by keeping up to date with just the English version and the data item.

I said all of this from my perspective as a Spanish speaker that always finds incongruences in articles in Spanish, but I can read English and French; however, this is not always the case for all Spanish speakers. Thus, if the documentation is wrong in Spanish, the wiki is transmitting some incorrect information to the mappers, and then the map will be affected. Data items could be the solution to keep the documentation updated without too much work.

**Comment by tomhughes** [comment]
We don't currently install from tar files, partly because that is a pain to automate, and partly because we were trying to track point releases automatically without needing a human to go in and bump the version number in chef.

So currently we install from git, but that does cause us a lot of additional pain on top of any wikibase issues as it's not really supported so it probably is time to change the way we handle those things.

The plan however, after discussing things with @nyurik on our ops call last week, was to try looking at dockerising things.

I find those release notes interesting for sure and I can see how they might have helped though I'm not sure I would have understood their importance as they read as if the php script is offering optimisation suggestions rather than that there are significant changes to how configuration works and that you need to run it to translate your current configuration to the new format - or at least I think that is what that script does?

Also it's useful to tell me federated properties won't work, except that I have no idea what they are or if we use them.

**Comment by 1ec5** [comment]
> Also it's useful to tell me federated properties won't work, except that I have no idea what they are or if we use them.

We aren’t using federated properties. The feature sounds like it would be useful for some sites out there, but it’s mutually exclusive of local properties, which is the whole point of the data items on the OSM Wiki.

**Comment by GreenReaper** [comment]
[Federated Properties are described in detail here](https://doc.wikimedia.org/Wikibase/master/php/docs_components_repo-federated-properties.html).

For what it's worth, significant work was done on a Federated Properties v2 which lets you add properties accessed via the API (internally, instead of just P1, it has the full URI to the remote entity). In this way you could have your local cake and eat it remotely, too.

[There is work to be done on it](https://phabricator.wikimedia.org/tag/wikibase_-_federated_properties/) and I think there was insufficient development bandwidth to complete this and roll out Wikibase.cloud at the same time.

Per comments accompanying the 1.37 release, "a new experimental Federated Properties version will be included in the next Wikibase Suite release, compatible with Mediawiki 1.38."

**Comment by matkoniecz** [comment]
@valeriewmde 

> We’d love to hear more about this. What exactly is being synced with what here?

In short: that is organisational issue caused by interface deficiencies and dislike of Wikibase (at least among some people)

Lets take https://wiki.openstreetmap.org/wiki/Tag:natural%3Dbirds_nest page

It has properties specified as infobox parameters in OSM Wiki

```
{{ValueDescription
|key=natural
|value=birds_nest
|image=File:Rufous_hornero_(Red_ovenbird)(Furnarius_rufus)_and_nest_(2).JPG
|description= A bird's nest.
|group=natural
|onNode=yes
|onWay=no
|onArea=yes
|onRelation=no
|seeAlso=
* {{Tag|man_made|nesting_site}}
|status=in use
}}
```

It also has [Q6219](https://wiki.openstreetmap.org/wiki/Item:Q6219) data item (not to be confused with [Q6219](https://www.wikidata.org/wiki/Q6219))

There were proposals to [eliminate infobox parameters and import all of them from Wikibase](https://wiki.openstreetmap.org/wiki/Talk:Wiki#Transition_to_use_data_items_when_this_can_be_done_without_loosing_information), but it is rejected by some as Wikibase interface is highly dysfunctional in many ways, watchlisting changes is basically impossible and it increases editing complexity in many ways (yes, it also reduces it in other ways and wiki editing has also some interface issues).

**Comment by matkoniecz** [comment]
@valeriewmde 

> >  broken watchlisting

> Please expand on this! We’d like to know more. What exactly is broken?

In case of wiki text I can watchlist https://wiki.openstreetmap.org/wiki/Tag:natural%3Dbirds_nest without watchlisting https://wiki.openstreetmap.org/wiki/Uk:Tag:natural%3Dbirds_nest

In case of watchlisting data item adding label in every single language will appear in my watchlist. That results in getting useless entries that I cannot verify `Added [uk] description: Гнізда птахів` may be correct, have slight mistake or be a slur - I lack language knowledge to judge it, so its appearance in my watchlist is useless.

I tried using data items and it caused my watchlist to become unusable. It become filled in 90% by edits in languages where I lack any ability to judge edit quality.

As result, data quality in data item descriptions is lower as watchlisting them is much harder and basically impossible on larger scale.

**Comment by matkoniecz** [comment]
> Thus, if the documentation is wrong in Spanish, the wiki is transmitting some incorrect information to the mappers, and then the map will be affected. Data items could be the solution to keep the documentation updated without too much work.

not really, data items can help at most with some infobox properties like "is it expected to be used on nodes" and "tag status (in use/de facto/deprecated)"

It is actively worse with translating tag descriptions and is making harder to handle longer description due to splitting editing between data item and wiki page

**Comment by GreenReaper** [comment]
Is there a way to get [client editing](https://mediawiki.org/wiki/Wikidata_Bridge) working on OSM Wiki? That might help with the "editing in two places".

[Wikidata users have made copious JavaScript tools](https://wikidata.org/wiki/Wikidata:Tools/Enhance_user_interface) to get around shortcomings in the official UI. I wonder if some of these might be adaptable to OSM Wiki. Stuff like WikidataWatchlistLabels and DiffLists might help.

**Comment by AndrewHain** [comment]
My comment to anyone who prefers the time before data items: this would take us back to the Verdy p era, with wiki work on the items starting just a week after he was blocked permanently. Over the past few years there has been a lot of effort to stop the bloat that he introduced from choking the wiki; some of that work uses data items and would be lost.

**Comment by matkoniecz** [comment]
I am confused why removing data items would require bringing back Verdy p or their inventions.

**Comment by 1ec5** [comment]
> Is there a way to get [client editing](https://mediawiki.org/wiki/Wikidata_Bridge) working on OSM? That might help with the "editing in two places".

Wikidata Bridge sounds like a good fit for our use case. Does it depend on Wikibase Client, or does it also allow editing items directly from within articles on wikis that integrate Wikibase directly? Back in 2019, @nyurik installed [WE-Framework](https://github.com/nyurik/WE-Framework/tree/osm) as [an optional gadget](https://wiki.openstreetmap.org/wiki/Data_items#Editing) on the wiki, but it was still a work in progress and apparently no longer works. If Wikibase Client is in better shape, I’d like to know what’s involved in setting it up. As an administrator, I can install any necessary changes to site scripts and protected templates, but extension-level changes would need the operations team’s involvement. Is it currently installed somewhere where we could try it out?

>> with wiki work on the items starting just a week after he was blocked permanently
> 
> As far as I know that was utterly unrelated - is it some attempt to credit data items for their ban? (even if their reaction to data items was a final straw, that is utterly unrelated to usefulness of data items)

Verdy_p was [blocked](https://wiki.openstreetmap.org/wiki/User_talk:Verdy_p#Blocked) for reasons unrelated to data items. I think @AndrewHain’s point is that data items have been an important tool for keeping the wiki’s templates manageable. The timing with Verdy_p was coincidental. (That said, Scribunto probably had a more direct impact on page performance. I sure hope we never uninstall Scribunto on the same rationale of upgrade complexity, because that really would take us back to the Stone Age.)

**Comment by dpriskorn** [comment]
> @valeriewmde 
> 
> 
> 
> > >  broken watchlisting
> 
> 
> 
> > Please expand on this! We’d like to know more. What exactly is broken?
> 
> 
> 
> In case of wiki text I can watchlist https://wiki.openstreetmap.org/wiki/Tag:natural%3Dbirds_nest without watchlisting https://wiki.openstreetmap.org/wiki/Uk:Tag:natural%3Dbirds_nest
> 
> 
> 
> In case of watchlisting data item adding label in every single language will appear in my watchlist. That results in getting useless entries that I cannot verify `Added [uk] description: Гнізда птахів` may be correct, have slight mistake or be a slur - I lack language knowledge to judge it, so its appearance in my watchlist is useless.
> 
> 
> 
> I tried using data items and it caused my watchlist to become unusable. It become filled in 90% by edits in languages where I lack any ability to judge edit quality.
> 
> 
> 
> As result, data quality in data item descriptions is lower as watchlisting them is much harder and basically impossible on larger scale.

FWIW I have the same issue in Wikidata. It would be nice to be able to filter based on language to avoid notification fatigue.

**Comment by dpriskorn** [comment]
> WikidataWatchlistLabels

@GreenReaper I had [trouble](https://www.wikidata.org/w/index.php?search=WikidataWatchlistLabels&title=Special%3ASearch&profile=advanced&fulltext=1&ns0=1&ns1=1&ns2=1&ns3=1&ns4=1&ns5=1&ns6=1&ns7=1&ns8=1&ns9=1&ns10=1&ns11=1&ns12=1&ns13=1&ns14=1&ns15=1&ns120=1&ns121=1&ns122=1&ns123=1&ns146=1&ns147=1&ns640=1&ns641=1&ns710=1&ns711=1&ns828=1&ns829=1&ns1198=1&ns1199=1&ns2300=1&ns2301=1&ns2302=1&ns2303=1) finding that one. Do you have a link?

**Comment by dpriskorn** [comment]
> DiffLists

[Source](https://www.wikidata.org/wiki/User:Yair_rand/DiffLists.js)

**Comment by GreenReaper** [comment]
Sure - it was on the page I linked at the bottom, but you need to do a text search, and Wikidata search requires quotes for it to show up - [here is a direct link](https://en.wikipedia.org/wiki/User:Evad37/WikidataWatchlistLabels).

**Comment by not-my-profile** [comment]
> broken watchlisting

The relevant Phabricator task appears to be [T43686](https://phabricator.wikimedia.org/T43686).

**Comment by fititnt** [comment]
> > I would strongly be opposed to losing the knowledge base built up in the data items.
>
> @AndrewHain Can you explain what this knowledge base is and how people can use it, to add to what @not-my-profile wrote above? This probably isn't best done here, but an OSM diary entry might be a good place for it.

On https://www.openstreetmap.org/user/fititnt/diary/400329 diary, I sort of tried what @SomeoneElseOSM proposed.

---

> <s>Fact: the Wiki people and the Wikibase people have near irreconcilable points of view about the structure of the same information.</s>

I was wrong here (at least in the context of OSM). For example, to my surprise, one of what I would called earlier _Wiki people_ actually had even suggestion in a older talk page about _Wikidata:WikiProject Ontology_ to fix semantic errors on Wikidata itself. Anyway, the complain that things are outdated is objectively true.

**Comment by ValorNaram** [comment]
@dpriskorn 

> FWIW I have the same issue in Wikidata. It would be nice to be able to filter based on language to avoid notification fatigue.

You can create email filters to filter unrelevant changes out. If `Added [uk] description` is generated by mediawiki and not written by users you can create a rule with all language codes you don't know to throw such emails away.

Pseudo filter:
`If content.contains(["[fr]", "[de]"]) then throw to /dev/null`

---

@nyurik and other sysadmins

**As a developer and data consumer I appreciate the work of you to keep Data Items alive because it can allow us to generate templates in the future and to properly document tags/keys.**

I want to build a user friendly way of dealing with tags/keys for non-OSMlers and in the longer term allow them to build presets e.g. for retail chains this can be useful to build a preset all their retailers can use to build a corporate info design. This can contribute to https://wiki.openstreetmap.org/wiki/Name_Suggestion_Index . For that I am in need of a system allowing me to query key/tag metadata.

**Comment by dpriskorn** [comment]
> @dpriskorn
> 
> > FWIW I have the same issue in Wikidata. It would be nice to be able to filter based on language to avoid notification fatigue.
> 
> You can create email filters to filter unrelevant changes out. If `Added [uk] description` is generated by mediawiki and not written by users you can create a rule with all language codes you don't know to throw such emails away.
> 
> Pseudo filter: `If content.contains(["[fr]", "[de]"]) then throw to /dev/null`

I'm not getting updates via email currently. I'm talking about the user interface here: https://www.wikidata.org/w/index.php?hidemyself=1&hidebots=1&hidepreviousrevisions=1&hidecategorization=1&tagfilter=wikidata-ui&limit=500&days=14&enhanced=1&title=Special:Watchlist&urlversion=2

**Comment by altilunium** [comment]
> > I would strongly be opposed to losing the knowledge base built up in the data items.
> 
> @AndrewHain Can you explain what this knowledge base is and how people can use it, to add to what @not-my-profile wrote above? This probably isn't best done here, but an OSM diary entry might be a good place for it.

It's prominently used in artificial general intelligence research. Translating the natural language question like "where's the nearest shop that sells X" requires a machine representation of common-sense knowledge. In short, converting "knowledge" stored as natural-language text from OSM Wiki into machine readable Wikibase format is a step for machine to understand  the meaning and rule behind OSM tags and concepts.

You can learn more about this topics (knowledge base, knowledge representation and reasoning, ontology) on this [OSM diary](https://www.openstreetmap.org/user/fititnt/diary/400329). It seems that several people are interested to work and conduct further research on this aspect. That's why the very existence of Wikibase extension is quite vital for them.

**Comment by SomeoneElseOSM** [comment]
@altilunium I did read that diary entry but somewhere around "Differences in how data is structured makes hard for RDF triplestores (like Wikidata) to store less structured content" I gave up - it didn't seem to be relevant to OSM at all.  Lots of people (even me) have spent time working out what mappers mean when they tag X as Y, but someone trying to impose a fixed hierarchy on OSM will simply get it wrong most of the time.  

If you think OSM's wiki data items are any sort of fit for the representation of data in OSM, you are being misled.  The OSM wiki is supposed to describe OSM features to humans.  Right now it [is broken](https://wiki.openstreetmap.org/wiki/Talk:Map_features#Issues) because of this abomination, and no one knows how to (or can be bothered to) fix it.

**Comment by ValorNaram** [comment]
@SomeoneElseOSM

> If you think OSM's wiki data items are any sort of fit for the representation of data in OSM, you are being misled. The OSM wiki is supposed to describe OSM features to humans. Right now it [is broken](https://wiki.openstreetmap.org/wiki/Talk:Map_features#Issues) because of this abomination, and no one knows how to (or can be bothered to) fix it.

It is good solution to represent the metadata of tags/keys once Data Items are an established part of the OSM universe. The use of the tags/keys in combination with location data to produce content is the actual product of OSM.

> Lots of people (even me) have spent time working out what mappers mean when they tag X as Y

This is the wrong way of contributing to OSM. This is as idiotic as pretending `phone` to be different than `contact:phone`. People must follow the documentation or their data get misinterpreted or not interpreted at all. This problem can be solved by a proper exchange between the OSM community and data consumers. Something I requested many times that we do that. Developers around OSM feel that mappers don't have much clue how their data is being interpreted or not.

>  I gave up - it didn't seem to be relevant to OSM at all.

**This is entirely not true.** It is absolutely relevant to OSM. It is about making our keys/tags understandable to machines and therefore making it easier to write dynamic applications which actually can handle human language. Actually @altilunium explained this use case very good in

> Translating the natural language question like "where's the nearest shop that sells X" requires a machine representation of common-sense knowledge.

The idea of the user is requesting something in a human natural language, the computer interpreting it and then answering back in human natural language or any other form non-techies can work with is pretty old and can be dated back to the early computer days. Now with AI and semantic data formats as schema.org and other initiatives like the Wikimedia project with Data Items accomplished this idea is coming true. Also intelligent search of app functionality in apps is contributing to that idea.

## A digital home assistant
Imagine the following: The wheeled user Alex gets a phone call from one of their college asking if there is any interest in going in a cafe after the work. As an wheeled user Alex knows that going out requires a bit more planning and unfortunately not all places are wheelchair friendly. Alex asks the caller to hang on, he asks his digital home assistant if the cafe is wheelchair friendly, as a toilet for the disabled or a public toilet for the disabled near by. The digital home assistant processes that question using voice recognition, OSM Data Items and the OSM databae to answer that question. In our case the answer is "yes" so Alex can go with them and does not need to ask them for a rearrangement because of his disability.

## Example of MS Office
See the new MS Office as an example. Microsoft developed an in-app search which allows you to query/search app functions in a human friendly way. Example: Once I needed to change the way the footer was build in MS PowerPoint. And I didn't know how to come to this setting which allows me to change this. So I used the in-app search and searched for "footer" and MS PowerPoint shew me the setting window I wanted to go in. **This is a nice way of working. As this concept evolves it will remove many difficulties non-techies have.** It's an effective way of problem solving: Just tell someone/something what you want and you get back the answer you seek.

**Comment by matkoniecz** [comment]
> > Lots of people (even me) have spent time working out what mappers mean when they tag X as Y

> This is the wrong way of contributing to OSM. 

It is absolutely necessary part of doing anything with OSM data, especially when doing something novel or writing an OSM editor.

> People must follow the documentation or their data get misinterpreted or not interpreted at all. 

Documentation must be written before it gets used. And to write it often you need to work out what mappers mean when they tag X as Y.

Either way it is quote offtopic on question "how we can get Wikibase maintainable by sysadmins or how we remove it if that is not achievable"

**Comment by matkoniecz** [comment]
> The idea of the user is requesting something in a human natural language, the computer interpreting it and then answering back in human natural language

Data items save here about 10 hours out of about 500 000 hour project here and are entirely replaceable by a simple python script scrapping OSM Wiki. And that is being wildly charitable here. It is entirely possible that download using data items will take the same effort as directly from infoboxes for zero net benefit. Especially as such magic box will need article texts anyway.

**Comment by mmd-osm** [comment]
@AndrewHain : can you shed some light on the recent [EWG protocol](https://wiki.osmfoundation.org/wiki/Working_Group_Minutes/EWG_2022-11-21), which discusses on option to turn data items into OSM relations?

Why isn't this discussion happening here on Github?

**Comment by frodrigo** [comment]
Just to let you know that Elastic now funds the Sophox server with $200/month.
https://opencollective.com/sophox

**Comment by AndrewHain** [comment]
@mmd-osm We had a brainstorming session during the group meeting.

**Comment by AndrewHain** [comment]
@mmd-osm We had a brainstorming session during the group meeting.

**Comment by fititnt** [comment]
Wow, I was not aware about the https://wiki.openstreetmap.org/wiki/User:Roland.olbricht/Data_Items_as_Relations, but even as a draft, I would fully support this! While for ontology-like use I even commented that full dumps tend to be helpful (maybe even editable like in GitHub repositories), one limitation would still be how users could edit translations, however the proposed approach not only allows this, but is with tools that already are near the actual use with data.

From https://wiki.osmfoundation.org/wiki/Working_Group_Minutes/EWG_2022-11-21, **unless Wikibase is know to be used for any other functionality of the Wiki itself** (@1ec5 might be best person to talk about!), the use cases for OpenStreetMap are so specific that I think is better to make this change without looking back. Even basic implementation is still better. Extra comments

Sadly, for today's EWG meeting, I will still have no extra diary or something on my drafts (see my post scriptum below). Anyway, let me comment 4 points quickly:

## 1. One example of strategy to prepare Wikibase dump to be imported

I don't have the full query right now (the fact of having so many languages makes it a bit copy and paste, but the initial test is here https://wiki.openstreetmap.org/wiki/User:EmericusPetro/sandbox/OpenStreetMap_Data_Items_util), but with command line from Apache Jena is possible to query the published database dump from here https://wiki.openstreetmap.org/dump/wikibase-rdf.ttl.gz into an CSV or TSV file with all the relevant fields.

From that, is how to convert tabular data into XML of changesets. Or the script could parte RDF directly.

Trivia: the default way Wikibase creates extra features (like ranks) make it more complex to query data than would be if we make id customized.

## 2. About extract RDF back from overpass

Is it possible to make the inverse procedure to recreate the original content: just overpass have some RDF output in a predictable URL (might be GET or POST). However the schema from Wikibase has some redundant statements that anyway would not be relevant outside).

However it would be necessary to also save on the relations of the original Ps and Qs on the original data. If this strategy stores data from other places, the other places would also need this.

Trivia: overpass actually does not need to implement SPARQL (RDF is always queryable). A lot of things can be done by overpass just pre-selecting data for client-side. Some tools even allow users to run queries to remote URLs without asking to download dumps first. **And overpass output is granted to always be as up to date as possible**.

## 3. The implementation "Data Items as relations" explains how to use OpenStreetMap Relations as if it was "virtual RDF triplestore" storage (and I liked it!)

In an ideal World, there would exist one or more namespaces for RDF-like data (this would require a huge table with 3, maybe a few more columns for data type. However relations allow "two columns", just need a way for parsers of XML/JSON formar know how to extract the Subject (that is the first column). However, the proposed strategy is very, very flexible. Maybe too much flexibility, however, might not even be a problem until more usage to define better organization.

While generic canonical data on XML could be exported as RDF/Turtle (so by default would generate a Relation as Relation), this very specific kind of storage would need an explicit different way for the user to know that is not just getting the relation in RDF, but unpacking the data as if the relation will be discarded. I'm calling this "virtual RDF triplestore" because depending on how you document for implementations to extract data, someone could store anything inside relations if they could change the Subject (the Predicate and Object already would be possible by default).

>> _**TODO: note to self, create an roundtrip example from XML Relation to A) Turtle as OSM Relation and B) Turtle as unpacked original data**_

By the way, (in case we document how to unpack to general data) I fully recommend the document to **not** allow tools to let users use Nodes and Ways to change the Subject (e.g. allow them to become a Subclass or property, validation rule, etc), not even as option. Even if not vandalism or poor done, points on map (like today sometimes points already are used for like continents) selected just because the user asked by bbox, but then the user on client-side allow inferencing, could be used to change behavior of region like not allowing "maxspeed>110". If such things are contained in Relations make easier to migrate to dedicated namespace (and before them, reduce likelihood of "unintended bugs" users start to use heavily)

## 4. The idea of allowing store RDF-like data actually could allow far more flexibility on the data model (which is something relevant to EWG later)

I think this proposal, even if basic, makes it a step to help tools understand the data in a formal, strict way. Discussions on this would take a lot of time, in particular because more people already using tools would need to check by themselves how this could help documentation and validation. However, things start by how to store data. Even if for now it is mostly to migrate Data Items, the same tools to extract back to RDF or to synchronize from other places could be used for many other things.

However, while I think we obviously should keep the IDs, there's some quirks (in special if redundant) that should not be imported that are very specific to Wikidata model, like the ranks, because they would force to create ANOTHER relation, with another ID, just to say "normal rank".

```turtle
## Part from https://wiki.openstreetmap.org/dump/wikibase-rdf.ttl.gz (to show redundant wikibase:BestRank, wikibase:Statement, etc
# The actuall list of prefixes is 28 items
@prefix wd: <https://wiki.openstreetmap.org/entity/> .
@prefix s: <https://wiki.openstreetmap.org/entity/statement/> .
@prefix wdt: <https://wiki.openstreetmap.org/prop/direct/> .
@prefix p: <https://wiki.openstreetmap.org/prop/> .


# Ugly ugly not worth copy as inspiration
wd:Q5 wdt:P3 wd:Q9 ;
    wdt:P31 "Area"@en ;
    p:P3 s:Q5-f322be79-4f24-df1b-e922-9d94caaad5a7 .

s:Q5-f322be79-4f24-df1b-e922-9d94caaad5a7 a wikibase:Statement,
        wikibase:BestRank ;
    wikibase:rank wikibase:NormalRank ;
    ps:P3 wd:Q9 .
```

Also, with an handcrafted, focused Ontology designed to be for OpenStreetMap, most of those Wikidata prefixes that are several URLs for the same Qs with different semantics are not relevant at all outside Wikidata. We can (and should) have prefixes like for node , way, relation (and anyone looking at the turtle versions will notice that they resemble Level0 editor!) and also prefixes for things that would be used in queries (like the geoSPARQL).

**While still need revision** (in special the namespace), an way could be encoded like this (the example here is mostly to show that is possible to nest values, without create explicitly blank nodes which would immediately make the RDF/Turtle more ugly than XML version):

```turtle
PREFIX osmm: <https://example.org/todo-meta/>
PREFIX osmnode: <https://www.openstreetmap.org/node/>
PREFIX osmt: <https://example.org/todo-tag/>
PREFIX osmway: <https://www.openstreetmap.org/way/>
PREFIX osmx: <https://example.org/todo-xref/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

# Note: the osmnode:(number) could also be defined here for their details.
#           however, in RDF format, dont matter if triples appear before or after.

osmway:100
    osmm:changeset 115876760 ;
    osmm:timestamp "2022-01-07T13:12:29+00:00"^^xsd:dateTime ;
    osmm:type "w" ;
    osmm:user "XXXVII" ;
    osmm:version 16 ;
    osmt:bicycle "use_sidepath" ;
    osmt:foot "use_sidepath" ;
    osmt:highway "secondary" ;
    osmt:junction "roundabout" ;
    osmt:maxspeed "100" ;
    osmt:ref "St 2069" ;
    osmt:surface "asphalt" ;
    osmt:tmc "DE:32282/32283" ;
    osmt:zone:traffic "DE:rural" ;
    # Note: Sophox uses osmm:has (but not sure how is the semantics)
    # https://www.w3.org/TR/turtle/#h3_collections
    osmx:hasnodes (
            osmnode:260904
            osmnode:456386831
            osmnode:260897
            osmnode:456386832
            osmnode:260898
            osmnode:456386833
            osmnode:185986175
            osmnode:456386834
            osmnode:260899
            osmnode:456386835
            osmnode:260900
            osmnode:456386836
            osmnode:260901
            osmnode:456386837
            osmnode:260902
            osmnode:456386838
            osmnode:1807516891
            osmnode:260903
            osmnode:456386839
            osmnode:260904
        ) ;
.
```

However, Wikidata (which is optimized for speed) will create the examples without using nesting non-orderer `[]` or ordered `()`. Also, near all prefixes are exactly like Sophox, but the the way of format the nodes and use of nesting make it a bit more easy for people which don't want to deal with why some lines end with `;` and others with `.`

That's it! I will take some time to comment more!


----

PS.: The reason I like this idea so much is that while not the same thing I was drafting by inspiration on how Wikibase (which is not a true triplestore, but scale very well) would actually means that the best approach would be the OpenStreetMap API, similar to how https://github.com/openstreetmap/openstreetmap-website/pull/2485 was added, be also able to output for RDF/Turtle (e.g. the outputs that today suffixing `.json` would work with suffixing `.ttl`). I still think this is worth it, but with all the discussion needed to define the schema and make sure it is good enough to work with other tools, even being very optimistic, it would take at least one year. And even then, the end result would still require load data by individual URLs one by one. So even the most basic export of triples from Overpass would immediately be very useful, without the need of external tools.

**Comment by Firefishy** [comment]
Wiki broke again today when upgrading to PHP 8. Details here: https://github.com/nyurik/OsmWikibase/issues/1

**Comment by GreenReaper** [comment]
Not really Wikibase's fault, though. [The MediaWikiPageNameNormalizer callback signature changed back in MW 1.37](https://doc.wikimedia.org/mediawiki-core/master/php/classMediaWiki_1_1Site_1_1MediaWikiPageNameNormalizer.html), while PHP 8 converted a warning (that had been raised beforehand as an issue) into an error. A lot of things can break with PHP 8.

The fix looks good for now; if the objective was to obtain the name at the end of the redirect it might not give the right result, but it presumably wouldn't have been doing that before, either.

**Comment by nyurik** [comment]
I merged the https://github.com/nyurik/OsmWikibase/pull/2 by @Firefishy so this should work now.  I will need to review it in more depth going forward to see if it should be handled differently

**Comment by Firefishy** [comment]
The Wikibase is now widely used and while we still have issues with it, it is being supported.
